File: code-library/13-string-manipulation/string-manipulation.R
# ==============================================================================
#  MODULE 13: STRING MANIPULATION WITH STRINGR
#  Dataset: CES 2020 (cleaned in Module 1) + simulated string variables
#
#  Purpose: Work with string data in R using stringr (part of tidyverse).
#  stringr functions are consistent: all take the string as the first argument
#  and all start with str_ — easy to remember and autocomplete.
#
#  Key stringr functions:
#    str_to_lower / str_to_upper / str_to_title  — case conversion
#    str_trim / str_squish                        — remove whitespace
#    str_length                                   — string length
#    str_sub                                      — extract substring
#    str_c / str_glue                             — concatenate strings
#    str_replace / str_replace_all                — find and replace
#    str_detect                                   — does string match pattern?
#    str_extract / str_extract_all                — extract matching substring
#    str_split                                    — split on delimiter
#    str_pad / str_trunc                          — pad or truncate
# ==============================================================================

library(tidyverse)

ces <- readRDS("CES2020_clean.rds")


# ==============================================================================
# SECTION 1: CREATE EXAMPLE STRING VARIABLES
# ==============================================================================

# Simulate messy state names as you'd get from a messy CSV join
ces <- ces |>
  mutate(
    state_name = case_when(
      state_fips == 12 ~ "  Florida  ",
      state_fips == 13 ~ "GEORGIA",
      state_fips == 36 ~ "New York",
      state_fips == 48 ~ "texas",
      state_fips ==  6 ~ "California ",
      state_fips == 42 ~ "Pennsylvania",
      state_fips == 39 ~ "ohio",
      state_fips == 26 ~ "MICHIGAN",
      state_fips == 37 ~ " North Carolina",
      state_fips ==  4 ~ "Arizona",
      TRUE             ~ NA_character_
    )
  )

# Simulate an open-ended text field
ces <- ces |>
  mutate(
    econ_text = case_when(
      row_number() == 1 ~ "Things are getting worse every year",
      row_number() == 2 ~ "I feel pretty secure financially",
      row_number() == 3 ~ "very worried about job loss",
      row_number() == 4 ~ "stable but not great",
      TRUE              ~ NA_character_
    )
  )


# ==============================================================================
# SECTION 2: BASIC STRING FUNCTIONS
# ==============================================================================

# Case conversion
ces |>
  select(state_name) |>
  filter(!is.na(state_name)) |>
  mutate(
    lower  = str_to_lower(state_name),
    upper  = str_to_upper(state_name),
    title  = str_to_title(state_name)
  ) |>
  distinct() |>
  head(10)

# Trim whitespace
ces |>
  filter(!is.na(state_name)) |>
  mutate(
    state_trimmed = str_trim(state_name),        # removes leading + trailing spaces
    state_squish  = str_squish(state_name)       # also collapses internal spaces
  ) |>
  select(state_name, state_trimmed) |>
  distinct()

# String length (after trimming)
ces |>
  filter(!is.na(state_name)) |>
  mutate(len = str_length(str_trim(state_name))) |>
  count(len)

# Clean pipeline: trim + title case
ces <- ces |>
  mutate(
    state_clean = str_to_title(str_trim(state_name))
  )


# ==============================================================================
# SECTION 3: EXTRACTING SUBSTRINGS
# ==============================================================================

# str_sub(string, start, end) extracts characters at positions start through end
# Negative values count from the end of the string

ces |>
  filter(!is.na(state_clean)) |>
  mutate(
    first5   = str_sub(state_clean, 1, 5),
    last5    = str_sub(state_clean, -5, -1),
    char3    = str_sub(state_clean, 3, 3)    # just the third character
  ) |>
  select(state_clean, first5, last5) |>
  distinct()

# Detect whether state name contains "New"
ces |>
  filter(!is.na(state_clean)) |>
  mutate(has_new = str_detect(state_clean, "New")) |>
  count(has_new)


# ==============================================================================
# SECTION 4: FIND AND REPLACE
# ==============================================================================

# str_replace(string, pattern, replacement) — replaces first match
# str_replace_all() — replaces all matches

ces |>
  filter(!is.na(state_clean)) |>
  mutate(
    # Remove all spaces
    no_spaces = str_replace_all(state_clean, " ", ""),
    # Standardize abbreviation
    fixed = str_replace(state_clean, "N\\. Carolina", "North Carolina")
  ) |>
  select(state_clean, no_spaces, fixed) |>
  distinct()


# ==============================================================================
# SECTION 5: REGULAR EXPRESSIONS
# ==============================================================================

# R uses PCRE regular expressions. The same special characters apply:
#   \\d   = digit (note double backslash in R strings)
#   \\s   = whitespace
#   \\w   = word character (letter, digit, underscore)
#   [0-9] = any digit
#   [A-Z] = any uppercase letter
#   ^     = start of string
#   $     = end of string
#   ()    = capture group (accessed via str_match or backreferences)
#   |     = OR

# Does state name contain any digit?
ces |>
  filter(!is.na(state_name)) |>
  mutate(has_digit = str_detect(state_name, "[0-9]")) |>
  filter(has_digit) |>
  nrow()   # should be 0

# Detect whether econ_text mentions job-related terms
ces |>
  filter(!is.na(econ_text)) |>
  mutate(
    mentions_job     = str_detect(str_to_lower(econ_text), "job|work|employ"),
    mentions_decline = str_detect(str_to_lower(econ_text), "worse|worried|loss|decline|bad")
  ) |>
  select(econ_text, mentions_job, mentions_decline)

# Extract 4-digit year from a date string
dates <- c("10/01/2020", "2020-10-02", "Oct 03, 2020", "10-04-2020")
str_extract(dates, "\\d{4}")   # extract first 4-digit sequence

# Extract month from "10/01/2020" format
str_match("10/01/2020", "(\\d{2})/(\\d{2})/(\\d{4})")
# Column 2 = month, column 3 = day, column 4 = year

# Validate email format
emails <- c("user@domain.com", "bademail", "user@.com", "good@uni.edu")
str_detect(emails, "^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$")


# ==============================================================================
# SECTION 6: SPLITTING STRINGS
# ==============================================================================

# str_split() returns a list (one element per input string)
# Use tidyr::separate() for splitting inside a dataframe — much more convenient

# Simulate a survey date variable
dates_df <- tibble(
  id   = 1:4,
  date_str = c("10/01/2020", "10/02/2020", "10/03/2020", "10/04/2020")
)

# separate() splits on a delimiter into multiple columns
dates_df |>
  separate(date_str, into = c("month", "day", "year"),
           sep = "/", convert = TRUE)

# Simulate a name variable (Last, First format)
names_df <- tibble(
  id   = 1:4,
  name = c("Smith, John A.", "jones, mary", "BROWN, JAMES", "Williams, Sue")
)

names_df |>
  separate(name, into = c("last_name", "first_name"),
           sep = ",", extra = "drop") |>
  mutate(
    last_name  = str_to_title(str_trim(last_name)),
    first_name = str_to_title(str_trim(first_name))
  )

# str_split_fixed() returns a matrix instead of a list — easier for two-part splits
str_split_fixed(names_df$name, ",", n = 2)


# ==============================================================================
# SECTION 7: COMBINING STRINGS
# ==============================================================================

# str_c() is the tidyverse equivalent of paste0() — concatenates strings
# str_glue() is even more readable: embed variables with {var}

ces |>
  filter(!is.na(state_fips)) |>
  mutate(
    # Zero-pad state FIPS to 2 digits, county to 3 digits
    state_fips_padded  = str_pad(state_fips, width = 2, pad = "0"),
    county_fips_fake   = sample(1:900, nrow(ces), replace = TRUE),
    county_fips_padded = str_pad(county_fips_fake, width = 3, pad = "0"),
    full_fips          = str_c(state_fips_padded, county_fips_padded)
  ) |>
  select(state_fips, state_fips_padded, county_fips_padded, full_fips) |>
  head(5)

# str_glue() for readable string interpolation
respondent_id <- 12345
wave          <- "post"
str_glue("Respondent {respondent_id} — wave: {wave}")


# ==============================================================================
# SECTION 8: STRING OPERATIONS ON MERGE KEYS
# ==============================================================================

# When merging on string keys, standardize both sides first:
# 1. str_trim()       — remove whitespace
# 2. str_to_lower()   — consistent case
# 3. str_squish()     — collapse internal spaces

# Simulate a lookup table
state_lookup <- tibble(
  state_key = c("florida", "georgia", "new york", "texas",
                "california", "pennsylvania", "ohio", "michigan",
                "north carolina", "arizona"),
  state_abbrev = c("FL","GA","NY","TX","CA","PA","OH","MI","NC","AZ")
)

ces_keyed <- ces |>
  filter(!is.na(state_name)) |>
  mutate(
    state_key = str_to_lower(str_squish(str_trim(state_name)))
  )

ces_with_abbrev <- ces_keyed |>
  left_join(state_lookup, by = "state_key")

ces_with_abbrev |>
  filter(!is.na(state_name)) |>
  count(state_abbrev, sort = TRUE)


# ==============================================================================
# SECTION 9: WORD COUNTS AND TEXT FEATURES
# ==============================================================================

# Count words using str_count with a word-boundary pattern
ces |>
  filter(!is.na(econ_text)) |>
  mutate(word_count = str_count(econ_text, "\\S+")) |>  # \\S+ = non-whitespace runs
  select(econ_text, word_count)

# Extract all words
ces |>
  filter(!is.na(econ_text)) |>
  mutate(words = str_extract_all(str_to_lower(econ_text), "\\w+")) |>
  select(econ_text, words) |>
  head(4)

# Frequency of specific terms
ces |>
  filter(!is.na(econ_text)) |>
  summarise(
    n_mentions_job     = sum(str_detect(str_to_lower(econ_text), "job|work")),
    n_mentions_decline = sum(str_detect(str_to_lower(econ_text), "worse|worried|loss"))
  )


# ==============================================================================
# SECTION 10: CLEANUP
# ==============================================================================

ces <- ces |>
  select(-state_name, -state_clean, -econ_text)

saveRDS(ces, "CES2020_clean.rds")
message("Module 13 complete.")

File: code-library/13-string-manipulation/string-manipulation.py
# ==============================================================================
#  MODULE 13: STRING MANIPULATION
#  Dataset: CES 2020 (cleaned in Module 1)
#
#  Stack: pandas (str accessor), re (regex), numpy
#  All string operations use the pd.Series.str accessor where possible.
# ==============================================================================

import pandas as pd
import numpy as np
import re

ces = pd.read_parquet("CES2020_clean.parquet")


# ==============================================================================
# SECTION 1: CASE CONVERSION
# ==============================================================================

# Simulate a messy name/label column
ces["race_label"] = ces["race_eth"].map({
    1: "White Non-Hispanic",
    2: "Black Non-Hispanic",
    3: "Hispanic",
    4: "asian non-hispanic",   # intentionally messy
    5: "MIXED RACE",
    6: "other",
    8: np.nan,
    9: np.nan,
})

ces["race_lower"] = ces["race_label"].str.lower()
ces["race_upper"] = ces["race_label"].str.upper()
ces["race_title"] = ces["race_label"].str.title()   # Title Case

print("Case conversions (first 6 distinct):")
print(
    ces[["race_label", "race_lower", "race_upper", "race_title"]]
    .drop_duplicates()
    .dropna()
    .to_string(index=False)
)


# ==============================================================================
# SECTION 2: TRIMMING AND PADDING
# ==============================================================================

# Simulate a state abbreviation column with extra whitespace
state_abbrev = pd.Series([" FL", "GA ", " TX ", "NY", "CA"])

state_clean = state_abbrev.str.strip()
print("\nBefore strip:", state_abbrev.tolist())
print("After strip: ", state_clean.tolist())

# str.strip()   — remove both ends
# str.lstrip()  — remove left only
# str.rstrip()  — remove right only

# Padding: zero-pad state FIPS codes to 2 digits
# pd.Series.str.zfill(width) pads with zeros on the left
fips_sample = pd.Series([1, 2, 4, 6, 12, 36, 48]).astype(str)
fips_padded = fips_sample.str.zfill(2)
print("\nFIPS padding:")
print(pd.DataFrame({"raw": fips_sample, "padded": fips_padded}).to_string(index=False))

# Left-justify with str.ljust() / right-justify with str.rjust()
labels = pd.Series(["age", "education", "ideology5"])
print("\nLeft-padded to width 15:")
print(labels.str.ljust(15, fillchar=".").tolist())


# ==============================================================================
# SECTION 3: SUBSTRINGS
# ==============================================================================

# str[start:stop] via str accessor — same slice syntax as Python strings
ces["race_first3"] = ces["race_label"].str[:3]

print("\nFirst 3 characters of race_label:")
print(ces[["race_label","race_first3"]].drop_duplicates().dropna().to_string(index=False))

# Extract the first word (everything before the first space)
ces["race_word1"] = ces["race_label"].str.split().str[0]

# str.find() equivalent: str.index — use str.find() returns -1 if not found
space_pos = ces["race_label"].str.find(" ")
print("\nPosition of first space in race_label (first 5 rows):")
print(space_pos.dropna().head().tolist())


# ==============================================================================
# SECTION 4: DETECTION  (str.contains / str.startswith / str.endswith)
# ==============================================================================

# str.contains(): returns True/False (NaN-safe with na=False)
ces["is_hispanic"] = ces["race_label"].str.contains("Hispanic", case=False, na=False)
ces["starts_w"]    = ces["race_label"].str.startswith("W", na=False)
ces["ends_ic"]     = ces["race_label"].str.endswith("ic", na=False)

print("\nDetection checks (first 6 distinct):")
print(
    ces[["race_label","is_hispanic","starts_w","ends_ic"]]
    .drop_duplicates()
    .dropna(subset=["race_label"])
    .to_string(index=False)
)

# Regex detection with str.match() (anchored at start) or str.contains(regex=True)
ces["is_mixed"] = ces["race_label"].str.contains(r"mixed|multi", case=False, na=False)


# ==============================================================================
# SECTION 5: REPLACE AND SUBSTITUTE
# ==============================================================================

# str.replace() — works with literal strings or regex
ces["race_clean"] = (
    ces["race_label"]
    .str.replace("Non-Hispanic", "NH", regex=False)
    .str.replace(r"\s+", " ", regex=True)   # collapse multiple spaces
    .str.strip()
)

print("\nAfter replace (first 6 distinct):")
print(
    ces[["race_label","race_clean"]]
    .drop_duplicates()
    .dropna()
    .to_string(index=False)
)

# str.replace with a function (callable)
# Example: capitalize only the first letter of each word
ces["race_sentence"] = ces["race_label"].str.lower().str.capitalize()


# ==============================================================================
# SECTION 6: SPLITTING AND JOINING
# ==============================================================================

# Simulate a "first last" name column
ces["full_name"] = "Respondent " + ces["caseid"].astype(str)

# str.split() returns a list per cell; use expand=True to get separate columns
name_split = ces["full_name"].str.split(" ", expand=True)
name_split.columns = ["name_part1", "name_part2"]

print("\nSplit names (first 5):")
print(name_split.head())

# Simulate a date string column (e.g., "2020-11-03")
ces["interview_date"] = "2020-11-" + ces["caseid"].mod(28).add(1).astype(str).str.zfill(2)

date_parts = ces["interview_date"].str.split("-", expand=True)
date_parts.columns = ["year", "month", "day"]

print("\nDate parts (first 5):")
print(date_parts.head())

# Joining: str.cat() concatenates series element-wise
state_fips_str = ces["state_fips"].astype(str).str.zfill(2)
ces["geo_id"] = "US" + state_fips_str


# ==============================================================================
# SECTION 7: REGEX EXTRACTION
# ==============================================================================

# str.extract() with named capture groups — returns a DataFrame of captures
sample_ids = pd.Series(["ID-001-FL", "ID-204-GA", "ID-089-TX", "ID-543-NY"])

extracted = sample_ids.str.extract(r"ID-(?P<num>\d+)-(?P<state>[A-Z]{2})")
print("\nRegex extraction:")
print(extracted)

# Extract just the numeric part from a single pattern
nums_only = sample_ids.str.extract(r"(\d+)", expand=False)
print("\nNumbers only:", nums_only.tolist())

# str.extractall() — all non-overlapping matches (useful for repeated patterns)
text_series = pd.Series(["age=30 sex=1", "age=45 sex=2", "age=22 sex=1"])
all_matches = text_series.str.extractall(r"(\w+)=(\d+)")
print("\nExtract all key=value pairs:")
print(all_matches.head(6))


# ==============================================================================
# SECTION 8: RE MODULE FOR COMPLEX PATTERNS
# ==============================================================================

# The re module is useful when you need lookaheads, named groups across rows,
# or operations that the str accessor doesn't support cleanly.

def extract_year(text):
    """Return the first 4-digit year found in a string, or None."""
    if pd.isna(text):
        return np.nan
    m = re.search(r"\b(19|20)\d{2}\b", text)
    return int(m.group()) if m else np.nan

sample_texts = pd.Series([
    "Survey conducted in November 2020",
    "Data collection: 2019–2020",
    "No year here",
    np.nan
])

years_found = sample_texts.apply(extract_year)
print("\nYear extraction with re.search:")
print(pd.DataFrame({"text": sample_texts, "year": years_found}).to_string(index=False))

# re.sub() equivalent for a scalar string
raw = "  CES  2020   Common   Content  "
clean = re.sub(r"\s+", " ", raw).strip()
print(f"\nre.sub collapse whitespace: '{clean}'")


# ==============================================================================
# SECTION 9: COUNTING CHARACTERS AND WORDS
# ==============================================================================

# Character length
ces["name_len"] = ces["full_name"].str.len()

# Count occurrences of a substring
text_sample = pd.Series([
    "strongly agree",
    "agree",
    "neither agree nor disagree",
    "disagree",
    "strongly disagree"
])

word_counts = text_sample.str.split().str.len()
agree_count = text_sample.str.count("agree")

print("\nWord/occurrence counts:")
print(pd.DataFrame({
    "text": text_sample,
    "n_words": word_counts,
    "n_agree": agree_count
}).to_string(index=False))


# ==============================================================================
# SECTION 10: STANDARDIZING MERGE KEYS
# ==============================================================================

# Real-world scenario: merge CES with an external state-level file where
# state names are inconsistently formatted.

ces_states = ces[["caseid", "state_fips"]].copy()

# Simulate a state file with messy names
state_info = pd.DataFrame({
    "state_name": [" Florida", "georgia", "TEXAS", "New York ", "California"],
    "state_fips": [12, 13, 48, 36, 6],
    "region_label": ["South", "South", "South", "Northeast", "West"]
})

# Standardize keys before merge
state_info["state_key"] = (
    state_info["state_name"]
    .str.strip()
    .str.lower()
    .str.replace(r"\s+", "_", regex=True)
)

print("\nStandardized state keys:")
print(state_info[["state_name","state_key","region_label"]].to_string(index=False))

# Merge on numeric FIPS (the safe approach when you have it)
ces_merged = ces_states.merge(
    state_info[["state_fips","region_label","state_key"]],
    on="state_fips",
    how="left"
)

print(f"\nMerge complete: {ces_merged['region_label'].notna().sum():,} matched rows")
print(ces_merged[["caseid","state_fips","region_label","state_key"]].head(8).to_string(index=False))


# ==============================================================================
# SECTION 11: BUILDING FIPS AND GEO CODES FROM PARTS
# ==============================================================================

# Construct a 5-digit county FIPS by zero-padding state (2) + county (3)
# Simulate county codes for demonstration
rng = np.random.default_rng(20240101)
ces["county_fips_raw"] = rng.integers(1, 999, size=len(ces))

state_str  = ces["state_fips"].astype(str).str.zfill(2)
county_str = ces["county_fips_raw"].astype(str).str.zfill(3)
ces["fips5"] = state_str + county_str

print("\nFIPS-5 construction (first 5):")
print(ces[["state_fips","county_fips_raw","fips5"]].head().to_string(index=False))


# ==============================================================================
# SECTION 12: ENCODING AND DECODING CATEGORICALS
# ==============================================================================

# pd.Categorical for memory-efficient storage of string variables
ces["race_cat"] = pd.Categorical(ces["race_title"].fillna("Unknown"))

print("\nRace categories:")
print(ces["race_cat"].value_counts())

# Map numeric codes back to labels for display
party_labels = {1: "Democrat", 2: "Republican", 3: "Independent"}
ces["party_label"] = ces["party_id3"].map(party_labels)

sex_labels = {1: "Male", 2: "Female"}
ces["sex_label"] = ces["sex"].map(sex_labels)

print("\nParty label distribution:")
print(ces["party_label"].value_counts(dropna=False))


# ==============================================================================
# SECTION 13: CLEANUP AND SAVE
# ==============================================================================

drop_cols = [
    "race_label", "race_lower", "race_upper", "race_title",
    "race_first3", "race_word1", "is_hispanic", "starts_w",
    "ends_ic", "is_mixed", "race_clean", "race_sentence",
    "full_name", "name_len", "interview_date", "geo_id",
    "county_fips_raw", "fips5", "race_cat", "party_label", "sex_label"
]

ces = ces.drop(columns=[c for c in drop_cols if c in ces.columns])

ces.to_parquet("CES2020_clean.parquet", index=False)
print("\nModule 13 complete.")
